pkg/client/introduce/client.go

```diff
@@ -0,0 +1,172 @@
+)
+
+// Provider contains dependencies for the introduce protocol and is typically created by using aries.Context()
+type Provider interface {
+	Service(id string) (interface{}, error)
+	StorageProvider() storage.Provider
+}
+
+// Client enable access to introduce API
+type Client struct {
+	service.Event
+	service service.DIDComm
+	store   storage.Store
+}
+
+// New return new instance of introduce client
+func New(ctx Provider) (*Client, error) {
+	svc, err := ctx.Service(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	introduceSvc, ok := svc.(service.DIDComm)
+	if !ok {
+		return nil, errors.New("cast service to Introduce Service failed")
+	}
+
+	store, err := ctx.StorageProvider().OpenStore(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		Event:   introduceSvc,
+		service: introduceSvc,
+		store:   store,
+	}, nil
+}
+
+// SendProposal sends proposal to introducees
+func (c *Client) SendProposal(dest1, dest2 *service.Destination) error {
+	return c.sendProposal(Options{
+		Destinations: []*service.Destination{dest1, dest2},
+	})
+}
+
+// SendProposalWithInvitation sends proposal to introducee
+// executes when the client has a public Invitation
+func (c *Client) SendProposalWithInvitation(inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(Options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+	})
+}
+
+// SendRequest sends request
+// sending a request means that introducee is willing to share its invitation
+// the function receives an invitation which can be sent to the introducer
+// an invitation will not be sent if introducer responded with a public invitation
+func (c *Client) SendRequest(invitation *didexchange.Invitation, destination *service.Destination) error {
+	return c.handleOutbound(&introduce.Request{
+		Type: introduce.RequestMsgType,
+		ID:   uuid.New().String(),
+	})
+}
+
+// HandleRequest executes after receiving a request (the client does not have a public Invitation)
+func (c *Client) HandleRequest(rID string, dest1, dest2 *service.Destination) error {
+	return c.sendProposal(Options{
+		Destinations: []*service.Destination{dest1, dest2},
+		ThreadID:     rID,
+	})
+}
+
+// HandleRequestWithInvitation executes after receiving a request when the client has a public Invitation
+func (c *Client) HandleRequestWithInvitation(rID string, inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(Options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+		ThreadID:     rID,
+	})
+}
+
+func (c *Client) handleOutbound(msg interface{}) error {
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	didMsg, err := service.NewDIDCommMsg(payload)
+	if err != nil {
+		return err
+	}
+
+	return c.service.HandleOutbound(didMsg, &service.Destination{})
+}
+
+// Options keeps the information needed for sending a proposal
+type Options struct {
```

Comment:
Is this planned to be used outside this package? (i.e., why exported?)

Suggest a more specific name in case you have need for more Option types.

Reply:
it is an internal thing, I am gonna make in unexported